lib/spack/spack/traverse.py

```diff
@@ -96,18 +96,80 @@ def __init__(self, visitor, key=id, visited=None):
         self.visited = set() if visited is None else visited
         self.key = key
 
-    def accept(self, node):
-        return self.visitor.accept(node)
+    def accept(self, item):
+        return self.visitor.accept(item)
 
-    def neighbors(self, node):
+    def neighbors(self, item):
         # Covering edges means: drop dependencies of visited nodes.
-        key = self.key(node.edge.spec)
+        key = self.key(item.edge.spec)
 
         if key in self.visited:
             return []
 
         self.visited.add(key)
-        return self.visitor.neighbors(node)
+        return self.visitor.neighbors(item)
+
+
+class TopoVisitor(object):
+    """Visitor that can be used in depth-first search to generate
+    a topological ordered list of vertices (we keep track of edges).
+
+    Algorithm based on "Section 22.4: Topological sort", Introduction to Algorithms
+    (2001, 2nd edition) by Cormen, Thomas H.; Leiserson, Charles E.; Rivest, Ronald L.;
+    Stein, Clifford.
+
+    The algorithm in plain English: realize that (1) we only append a vertex A before
+    traversing all possible paths to any descendant B (that means A is always appended
+    to reverse_order after all its descendants) and (2) while following paths from
+    A -> B, we push B to reverse_order the first time it is reached, and guarantee its
+    ancestors are always appended and its descendants are not (on the first path
+    this is thanks to backtracking, on any other path we do not follow B as its
+    already seen.)
+
+    The neat thing about this algorithm is that it can operate on an immutable DAG,
+    it doesn't require marking or popping edges, and it works in linear time (in
+    vertices and edges).
```

Comment:
Hm yeah, let me just drop it. It was in contrast to the Kahn algorithm where it's common delete edges, but even that can run in linear time if you keep a dict of `node id` -> `# in-edges left`.

Btw, notice that if we would do Kahn instead of this DFS/backtracking algorithm, we would still need to do one pass with BFS/DFS to actually get the `#in-edge` count of the sub-DAG we're traversing over. So I think this DFS algorithm is typically more efficient.